service/src/test/java/bio/terra/policy/service/policy/PolicyDataTrackingConstraintTest.java

```diff
@@ -0,0 +1,272 @@
+    assertTrue(dataTypeSet.containsAll(dataTypes));
+  }
+
+  @Test
+  void dataTrackingConstraintTest_combineWhenSourceHasFewerDataTypes() throws Exception {
+    var dataTrackingConstraint = new PolicyDataTrackingConstraint();
+
+    String dataType2 = DATA_TYPE_NAME + "2";
+    String dataType3 = DATA_TYPE_NAME + "3";
+
+    Set<String> dataTypes = new HashSet<>(Arrays.asList(DATA_TYPE_NAME, dataType2, dataType3));
+
+    var dependentPolicy =
+        new PolicyInput(
+            TERRA_NAMESPACE, DATA_TRACKING_CONSTRAINT, buildMultimap(DATA_TRACKING_KEY, dataTypes));
+
+    dataTypes.remove(dataType3);
+    var sourcePolicy =
+        new PolicyInput(
+            TERRA_NAMESPACE, DATA_TRACKING_CONSTRAINT, buildMultimap(DATA_TRACKING_KEY, dataTypes));
+
+    PolicyInput resultPolicy = dataTrackingConstraint.combine(dependentPolicy, sourcePolicy);
+    Set<String> dataTypeSet =
+        dataTrackingConstraint.dataToSet(resultPolicy.getAdditionalData().get(DATA_TRACKING_KEY));
+
+    assertEquals(3, dataTypeSet.size());
+    assertTrue(dataTypeSet.containsAll(dataTypes));
+  }
+
+  @Test
+  void dataTrackingConstraintTest_removeDataType() throws Exception {
+    var dataTrackingConstraint = new PolicyDataTrackingConstraint();
+
+    String dataType2 = DATA_TYPE_NAME + "2";
+    String dataType3 = DATA_TYPE_NAME + "3";
+
+    Set<String> dataTypes = new HashSet<>(Arrays.asList(DATA_TYPE_NAME));
+    var removePolicy =
+        new PolicyInput(
+            TERRA_NAMESPACE, DATA_TRACKING_CONSTRAINT, buildMultimap(DATA_TRACKING_KEY, dataTypes));
+
+    dataTypes.add(dataType2);
+    dataTypes.add(dataType3);
+    var targetPolicy =
+        new PolicyInput(
+            TERRA_NAMESPACE, DATA_TRACKING_CONSTRAINT, buildMultimap(DATA_TRACKING_KEY, dataTypes));
+
+    PolicyInput resultPolicy = dataTrackingConstraint.remove(targetPolicy, removePolicy);
+    Set<String> dataTypeSet =
+        dataTrackingConstraint.dataToSet(resultPolicy.getAdditionalData().get(DATA_TRACKING_KEY));
+
+    assertEquals(2, dataTypeSet.size());
+    assertTrue(dataTypeSet.containsAll(Arrays.asList(dataType2, dataType3)));
+  }
+
+  @Test
+  void dataTrackingConstraintTest_removeAllDataTypes() throws Exception {
+    var dataTrackingConstraint = new PolicyDataTrackingConstraint();
+
+    String dataType2 = DATA_TYPE_NAME + "2";
+    String dataType3 = DATA_TYPE_NAME + "3";
+
+    Set<String> dataTypes = new HashSet<>(Arrays.asList(DATA_TYPE_NAME, dataType2, dataType3));
+    var removePolicy =
+        new PolicyInput(
+            TERRA_NAMESPACE, DATA_TRACKING_CONSTRAINT, buildMultimap(DATA_TRACKING_KEY, dataTypes));
+
+    var targetPolicy =
+        new PolicyInput(
+            TERRA_NAMESPACE, DATA_TRACKING_CONSTRAINT, buildMultimap(DATA_TRACKING_KEY, dataTypes));
+
+    PolicyInput resultPolicy = dataTrackingConstraint.remove(targetPolicy, removePolicy);
+    assertNull(resultPolicy);
+  }
+
+  @Test
+  void dataTrackingConstraintTest_validation() {
```

Comment:
I think we'll want a test around ensuring that the validity check bombs out if a data type is not passed.